src/qibojit/custom_operators/backends.py

```diff
@@ -384,3 +383,252 @@ def swap_pieces(self, piece0, piece1, new_global, nlocal):
+                 workspaceSize
+                 )
+        self.cusv.apply_matrix(*args2)
+        self.cusv.destroy(handle)
+        return state
+
+    def multi_qubit_base(self, state, nqubits, targets, qubits=None, gate=None):
+        handle = self.cusv.create()
+        state = self.cast(state)
+        ntarget = len(targets)
+        if qubits is None:
+            qubits = self.cast(sorted(nqubits - q - 1 for q in targets), dtype = self.cp.int32)
+        target = [ nqubits - q - 1 for q in targets]
+        target = self.np.asarray(target[::-1], dtype = self.np.int32)
+        controls = self.np.asarray([i for i in qubits.get() if i not in target], dtype = self.np.int32)
+        ncontrols = len(controls) if qubits is not None else 0
+        adjoint = 0
+        gate = self.cast(gate)
+        assert state.dtype == gate.dtype
+        data_type, compute_type = self.get_cuda_type(state.dtype)
+
+        if isinstance(gate, self.cp.ndarray):
+            gate_ptr = gate.data.ptr
+        elif isinstance(gate, self.np.ndarray):
+            gate_ptr = gate.ctypes.data
+        else:
+            raise ValueError
+
+        args1 = (handle,
+                 data_type,
+                 nqubits,
+                 gate_ptr,
+                 data_type,
+                 self.cusv.MatrixLayout.ROW,
+                 adjoint,
+                 ntarget,
+                 ncontrols,
+                 compute_type
+                 )
+        workspaceSize = self.cusv.apply_matrix_buffer_size(*args1)
+
+        # check the size of external workspace
+        if workspaceSize > 0:
+            workspace = self.cp.cuda.memory.alloc(workspaceSize)
+            workspace_ptr = workspace.ptr
+        else:
+            workspace_ptr = 0
+
+        args2 = (handle,
+                 state.data.ptr,
+                 data_type,
+                 nqubits,
+                 gate_ptr,
+                 data_type,
+                 self.cusv.MatrixLayout.ROW,
+                 adjoint,
+                 target.ctypes.data,
+                 ntarget,
+                 controls.ctypes.data,
+                 ncontrols,
+                 0,
+                 compute_type,
+                 workspace_ptr,
+                 workspaceSize
+                 )
+        self.cusv.apply_matrix(*args2)
+        self.cusv.destroy(handle)
+        return state
+
+    def transpose_state(self, pieces, state, nqubits, order):
+        raise NotImplementedError("`transpose_state` method is not "
+                                  "implemented for GPU.")
+
+    def swap_pieces(self, piece0, piece1, new_global, nlocal):
+        raise NotImplementedError("`swap_pieces` method is not "
+                                  "implemented for GPU.")
+
+    def measure_frequencies(self, frequencies, probs, nshots, nqubits, seed=1234, nthreads=None):
+        raise NotImplementedError("`measure_frequencies` method is not "
+                                  "implemented for GPU.")
```

Comment:
These three methods are inherited from `CupyBackend` so it should be fine to remove them from here.

Reply:
Yes, you are right. I will remove these methods.